### ISSUE TITLE
[SPIKE] Revisit Removal of "Unit"

### ISSUE THREAD

**Issue by seanluyk** [ualbertalib/avalon#537]
### Descriptive summary
Initially, we thought that Unit wasn't a very useful feature, so Chris commented the facet out (see #17). However, as the service has grown, and we have groups with multiple collections, we're seeing the need for Unit as a way to subdivide collections, similarly to how Jupiter uses communities/collections. The metadata team is also interested in the possibility of using unit to allow for more collection-specific controlled vocabularies and validation (functionality still to be investigated). 

This spike is to investigate whether we could enable units, and how much work this would be.     

### Related work
#17, #305, #282   
 - [Managing Controlled Vocabularies](https://wiki.dlib.indiana.edu/display/VarVideo/Managing+Controlled+Vocabularies)
 - Defined [here](https://github.com/ualbertalib/avalon/blob/master/config/controlled_vocabulary.yml)

**Comment by seanluyk** [comment]
@weiweishi @jefferya I'd like to elevate the importance of this issue and perhaps see if it could be worked on in an upcoming sprint. As outlined above, with more collections coming in from individual creators, unit is becoming much more needed as a way to subdivide collections. I believe that you'll need a list of units to add to a config file, which @anayram and I can provide soon

**Comment by anayram** [comment]
Two units that come to mind that we could try to get started and/or for testing controlled vocabularies:

Kule Folklore Centre
University of Alberta Libraries

**Comment by seanluyk** [comment]
@anayram I agree, and I've begun an analysis of the various collections and how they might be amalgamated in this Trello card as well FYI: https://trello.com/c/zJBqChGu

**Comment by jefferya** [comment]
@seanluyk @anayram 
This if straightforward to add back. A couple of questions:
* is there an updated list of units? 
* this will need to be decomposed into multiple steps
  1. code change and initial vocabulary (contains `default unit` plus new unit i.e., interim step)
  2. cleanup - assuming the `default unit` will be changed
  3. `controlled_vocabulary.yml` update to remove `default unit`

**Comment by anayram** [comment]
@jefferya I believe you can use the list of units @seanluyk recommends here: https://trello.com/c/zJBqChGu

**Comment by jefferya** [comment]
Note: To test via UI, change `unit` during edit/creation of a collection. The `unit` change is then reflected in the members of the collection (e.g., the following API call will show the details: http://localhost:3000/media_objects/:id.json)

![Screenshot from 2020-04-14 15-02-18](https://user-images.githubusercontent.com/879785/79274144-59199e00-7e61-11ea-894b-f4639f9deee2.png)

**Comment by seanluyk** [comment]
@anayram @jefferya I'd say that the list in Trello isn't fully fleshed out yet, but Mariana and I can come up with an updated list of units a.s.a.p. I'll also provide you with access to our repositories Trello board, Jeff

**Comment by jefferya** [comment]
@seanluyk @anayram @
Can we pause the "make sure Unix and GUI names match" requirement in <https://trello.com/c/zJBqChGu>? The reasoning, changing the Unix filesystem directory name will cause some challenges for preservation if not carefully done. 
Background:
Avalon records the directory it expects to find the masterfile and this file_location property doesn't get updated and may make preservation harder in the following cases:
* Batch dropbox directory created/renamed via sftp or Unix/Linux command line doesn't update Avalon masterfile properties or collection properties
* Web UI collection rename: doesn't appear to change the underlying Unix/Linux filesystem directory name (e.g., change "Test Collection" to "Sandbox" and the underlying dropbox directory naem remains "Test_Collection" (there's a spreadsheet of these disconnections between collection name and directory name)
* creation of a new collection via the Web UI and there is a collision of names (e.g., multiple Test Collection yields "Test_Collection_02", I think)  

In summary, care must be take to avoid creating a disconnect between where Avalon thinks a masterfile exists and where the masterfile is located on the Unix/Linux filesystem. I think ideas on how best implement the ideas in the Trello card will flow from discussions related to implementing preservation.

**Comment by seanluyk** [comment]
@jefferya absolutely, thanks for bringing this up. This is really just on our list as something to eventually clean up, so can wait until our preservation solution is in place to make any changes

**Comment by anayram** [comment]
@seanluyk sounds good. I created a quick listing of collections and associated units in case it helps to get this ready for @jefferya: https://docs.google.com/spreadsheets/d/198c8Tkh1ru2fsLusfCp66Y5FHNSRsL7obxj3Hwg_Sns

Items highlighted have identical unit / collection names and might need some additional grouping.

**Comment by jefferya** [comment]
* Thank you for the doc @anayram -- one question, for the rows in the spreadsheet that have a blank unit, for example "Convocation Hall", what should associated Avalon Unit be?  

Should we chat during the next ERA A+V meeting?

**Comment by seanluyk** [comment]
Thanks for putting this together, @anayram! I'm adding some ideas in the blank rows, and agree with @jefferya that we could discuss this at our next team meeting. Many of the collection names will need to be changed as a result, which is OK I think, and we can start to document those changes as well.

**Comment by seanluyk** [comment]
@jefferya it also just occurred to me that Chris had removed the Unit facet from the UI as well. Is that significant to add back?

**Comment by jefferya** [comment]
Adding back the Unit facet (restoring to upstream) is trivial, the details are in this work in progress (WIP) PR for reference: <https://github.com/ualbertalib/avalon/pull/634/files>

Todo: I haven't fully tested unit across the system (tested manually: facet appears, search limits by unit, unit can be set at collection level during create/edit, unit appears in item JSON metadata output, and unit appears on the media object UI view).

**Comment by anayram** [comment]
Thanks @seanluyk @jefferya.  Sounds good to me. One thing I'd like to check is how to add different vocabularies to different units now that they are available but I imagine this is lower priority.

**Comment by seanluyk** [comment]
@anayram I don't believe that units allow for any additional functionality related to metadata, permissions, etc. although that would be useful. I could be wrong. Afaik, units just provide us with an additional cross-cutting facet, so are sort of like Communities in ERA

**Comment by anayram** [comment]
@seanluyk I think you are right, there are no additional functionalities
for units. I must have misinterpreted the documentation. Thank you for the
clarification!

**Comment by jefferya** [comment]
Todo:
- investigate how to change unit on existing collection members. Preexisting collection, edit collection and change unit  to 'y' -- the result, members of this collection do not appear in the facet 'y' (or the facet 'y' doesn't appear in the list). See Sept 3rd e-mail chain.

**Comment by jefferya** [comment]
Further diagnosis

* a unit change for a collection triggers a background job that is queued at a later time
* depending on the number of jobs in the queue and workers, the unit change may not instantaneously update the members of the collection
* also, the unit facet is similarly affected

Reference:
* https://github.com/ualbertalib/avalon/blob/b7006cee4ac6f03266b91741052b65bda5fb9918/app/models/admin/collection.rb#L57
* https://github.com/ualbertalib/avalon/blob/b7006cee4ac6f03266b91741052b65bda5fb9918/app/models/admin/collection.rb#L155-L158
